book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -165,49 +227,232 @@ each rooted at the highest finalized block. Each chain consists of a map from
-  another chain, calls push repeatedly
-- **ChainSet**: `(BTreeSet<Chain>, BTreeMap<block::Height, Arc<Block>>)`
-  - `fn finalize(&mut self) -> Arc<Block>`
-  - `fn commit_block(&mut self, block: Arc<Block>) -> Result<(), Err>`
-    - iterate through chains, if the parent of `block` is the tip of any
-    chains, try to push block onto that chain
-    - if not, iterate through chains, checking if the parent is contained in
-    that chain
-    - if so, fork the chain at the parent and try to push `block` onto that
-    fork and add newly extended chain if the push succeeds
-    - if not, queue block, prune queued blocks that are below the reorg limit
-
-
-- XXX work out whether we should store extra data (e.g., a HashSet of UTXOs
-  spent by some block etc.) to speed up checks.
-
-When a new block extends the best chain past 100 blocks, the old root is
-removed from the non-finalized state and committed to the finalized state.
+### `Chain` Type
+[chain-type]: #chain-type
 
-## Committing non-finalized blocks
+The `Chain` type consists of a set of blocks, containing the non-finalized
+portion of the chain it represents where the lowest height block's parent is
+the tip of the finalized state. All of the other members of `Chain` cache
+information contained within that set of blocks for fast lookup.
 
-If the parent block is not committed, add the block to an internal queue for
-future processing.
+We represent the non-finalized portion of a chain with the following data
+structure and API:
+
+```rust
+struct Chain {
+    blocks: BTreeMap<block::Height, Arc<Block>>,
+    height_by_hash: HashMap<block::Hash, block::Height>,
+    tx_by_hash: HashMap<transaction::Hash, (block::Height, tx_index)>,
+
+    utxos: HashSet<transparent::Output>,
+    sapling_anchors: HashSet<sapling::tree::Root>,
+    sprout_anchors: HashSet<sprout::tree::Root>,
+    sapling_nullifiers: HashSet<sapling::Nullifier>,
+    sprout_nullifiers: HashSet<sprout::Nullifier>,
+    partial_cumulative_work: PartialCumulativeWork,
+}
+```
+
+#### `pub fn push(&mut self, block: Arc<Block>) -> Result<(), Error>`
+
+Push a block into a chain as the new tip if the block is a valid extension of
+that chain.
+
+1. Run contextual validation checks on block against Self
+
+1. Update cummulative data members
+    - Add block to end of `self.blocks`
+    - Add hash to `height_by_hash`
+    - for each `transaction` in `block`
+      - add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+    - Add new utxos and remove consumed utxos from `self.utxos`
+    - Add anchors to the appropriate `self.<version>_anchors`
+    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add work to `self.partial_cumulative_work`
+
+#### `pub fn pop_root(&mut self) -> Arc<Block>`
+
+Remove the lowest height block of the non-finalized portion of a chain.
+
+1. Remove the lowest height block from `self.blocks`
+
+1. Update cummulative data members
```

Comment:
Here and elsewhere the numbered lists use 1. 1. 1. 1., etc. to do auto-numbering.  I think this is fine for writing the document but it makes it harder to read as text, so ideally before merging this we could replace them with correct numbers in the text file.